src/axom/core/Map.hpp

```diff
@@ -0,0 +1,739 @@
+  {
+    init(num_buckets, bucket_len);
+    m_end.key = Key {0};
+    m_end.value = T {0};
+    m_end.next = -2;
+  }
+  /// @}
+
+  /*!
+   * Destructor. Frees associated buffers.
+   */
+  ~Map() { deallocate(); }
+
+  /// \name Map Methods
+  ///@{
+  /*!
+   * \brief Resizes Map by creating new Map instance and inserting all values
+   * from original Map.
+   *
+   * \param [in] buckets user-specified number of buckets in new Map. -1 to fall back to multiplicative factor.
+   * \param [in] factor user-specified multiplicative factor to determine size of new Map based upon existing. 
+   *
+   * \return true if the hash table is now in a safe state, false otherwise
+   *
+   * \note if neither input is specified, the new Map will be of size 2*old Map size. 
+   * \note both map instances exist in memory until rehash returns. Risks running out of memory.
+   * \note Rehash should generally ensure a previously filled bucket is no longer filled, and the load_factor is correct,
+   *  if used properly. However, since this highly manual process has room for error this implementation does
+   *  perform a check for if any bucket is now full. 
+   */
+  bool rehash(int buckets = -1, int factor = -1)
+  {
+    int newlen = 0;
+    IndexType ind;
+    std::size_t hashed;
+    if(buckets != -1 && buckets > m_bucket_count)
+    {
+      newlen = buckets;
+    }
+    else if(factor != -1 && factor > 1)
+    {
+      newlen = factor * m_bucket_count;
+    }
+    else
+    {
+      newlen = 2 * m_bucket_count;
+    }
+    m_bucket_fill = false;
+    axom_map::Bucket<Key, T>* new_list = alloc_map(newlen, m_bucket_len);
+
+    for(int i = 0; i < m_bucket_count; i++)
+    {
+      ind = m_buckets[i].m_head;
+      while(ind != -1)
+      {
+        hashed = (get_hash(m_buckets[i].m_list[ind].key) % newlen);
+        new_list[hashed].insert_no_update(m_buckets[i].m_list[ind].key,
+                                          m_buckets[i].m_list[ind].value);
+        ind = m_buckets[i].m_list[ind].next;
+      }
+      axom::deallocate(m_buckets[i].m_list);
+    }
+    axom::deallocate(m_buckets);
+    m_buckets = new_list;
+    m_bucket_count = newlen;
+    for(int i = 0; i < m_bucket_count; i++)
+    {
+      if(m_buckets[i].get_size() == m_buckets[i].get_capacity())
+      {
+        m_bucket_fill = true;
+      }
+    }
+    return check_rehash();
+  }
+  /*!
+   * \brief Deallocates memory resources for this Map instance, resets state as if constructed with 0 elements.
+   *  Also used in destructor.
+   *
+   */
+  void deallocate()
+  {
+    for(int i = 0; i < m_bucket_count; i++)
+    {
+      axom::deallocate(m_buckets[i].m_list);
+    }
+    axom::deallocate(m_buckets);
+    m_bucket_count = 0;
+    m_bucket_len = 0;
+    m_size = 0;
+    m_load_factor = 0;
+  }
+
+  /*!
+   * \brief Initializes this Map instance, mostly in case deallocate was used and there's a desire to continue use.
+   *
+   */
+  void init(int num_buckets, int bucket_len = 10)
```

Comment:
Is this required before the user adds any new data into a Map that has been deallocated?  What happens if they deallocate, do not call this, and then try an insert?

Reply:
I don't think I've put any amount of protection on for that. I could put a conditional somewhere, perhaps, or better document to not clear then insert?

Reply:
Definitely make it clear in the documentation that it is required.  I think if someone tries to insert with zero buckets it could be an ugly segfault, so it may be good to have some error checking there as well.

Reply:
I got that in, though I'm not necessarily pleased with the outcome -- may just revert to documentation, and throw an assert at it for debug builds.